### PR TITLE
Add an integration test for CNI race condition repair.

### PR DESCRIPTION
**Pull request by bianpengyuan**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: tests/integration/pilot/cni/cni_race_test.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+func TestCNIRaceRepair(t *testing.T) {
+	framework.NewTest(t).
+		Features("traffic.cni.race-condition-repair").
+		Run(func(t framework.TestContext) {
+			if !ist.Settings().EnableCNI {
+				t.Skip("CNI race condition mitigation is only tested when CNI is enabled.")
+			}
+			cluster := t.Clusters().Default()
+			// To begin with, delete CNI Daemonset to simulate a CNI race condition.
```
Comment (howardjohn) [inline tests/integration/pilot/cni/cni_race_test.go:66]: Why is this enough? Once its installed once isn't that sufficient to get iptables? Or is it cleaning up on shutdown

File: tests/integration/pilot/cni/cni_race_test.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+			deleteCNIDaemonset(t, cluster)
+
+			// Rollout restart instances in namespace 1, and wait for a broken instance.
+			rolloutCmd := fmt.Sprintf("kubectl rollout restart deployment -n %s", apps.Namespace1.Name())
```
Comment (howardjohn) [inline tests/integration/pilot/cni/cni_race_test.go:70]: There is a .Restart() on echo.Instance that may be preferred.

File: tests/integration/pilot/cni/cni_race_test.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+		Features("traffic.cni.race-condition-repair").
+		Run(func(t framework.TestContext) {
+			if !ist.Settings().EnableCNI {
+				t.Skip("CNI race condition mitigation is only tested when CNI is enabled.")
+			}
+			cluster := t.Clusters().Default()
+			// To begin with, delete CNI Daemonset to simulate a CNI race condition.
+			deleteCNIDaemonset(t, cluster)
+
+			// Rollout restart instances in namespace 1, and wait for a broken instance.
```
Comment (howardjohn) [inline tests/integration/pilot/cni/cni_race_test.go:69]: Do we need all? Or just a single pod?

File: tests/integration/pilot/cni/cni_race_test.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+			return errors.New("CNI Daemonset pod still exists after deletion")
+		}
+		return nil
+	}, retry.Delay(1*time.Second), retry.Timeout(80*time.Second))
+}
+
+func deployCNIDaemonset(ctx framework.TestContext, c cluster.Cluster) error {
+	args := []string{
+		"install", "-f", filepath.Join(env.IstioSrc,
+			"tests/integration/pilot/cni/testdata/cni.yaml"),
```
Comment (howardjohn) [inline tests/integration/pilot/cni/cni_race_test.go:108]: can we save the old yaml and re-apply it? There may be customizations done outside the test framework?

File: tests/integration/pilot/cni/cni_race_test.go (hunk 1 above)
Comment (bianpengyuan) [reply tests/integration/pilot/cni/cni_race_test.go:66]: Yeah, CNI network plugin gets cleaned up on removal.

File: tests/integration/pilot/cni/cni_race_test.go (hunk 2 above)
Comment (bianpengyuan) [reply tests/integration/pilot/cni/cni_race_test.go:70]: Tricky thing is that workload will never become ready until CNI gets reinstalled. However `Restart` waits for workload to be ready, and I don't want to add a SkipWaiting to `Restart` just for this test. Let me know if you think it is fine to add skipWaiting option.

File: tests/integration/pilot/cni/cni_race_test.go (hunk 3 above)
Comment (bianpengyuan) [reply tests/integration/pilot/cni/cni_race_test.go:69]: I think just a single pod is enough, which indicates there are pods stuck in race condition. Later we verify that no pod is broken, which proves that repair has kicked in.

File: tests/integration/pilot/cni/cni_race_test.go (hunk 4 above)
Comment (bianpengyuan) [reply tests/integration/pilot/cni/cni_race_test.go]: Done.

File: tests/integration/pilot/cni/cni_race_test.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	"istio.io/istio/tests/integration/security/util"
+	"istio.io/istio/tools/istio-iptables/pkg/constants"
+)
+
+var (
+	ist  istio.Instance
+	apps = &util.EchoDeployments{}
+)
+
+func TestMain(m *testing.M) {
```
Comment (howardjohn) [inline tests/integration/pilot/cni/cni_race_test.go:46]: why do we need a whole new test suite for this? It doesn't seem like its installing anything custom

File: tests/integration/pilot/cni/cni_race_test.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	framework.NewSuite(m).
+		Label(label.CustomSetup).
+		Setup(istio.Setup(&ist, nil)).
+		Setup(func(t resource.Context) error {
+			return util.SetupApps(t, ist, apps, false)
```
Comment (howardjohn) [inline tests/integration/pilot/cni/cni_race_test.go:51]: Why do we need so many? Don't we only need 1 to test this?

File: tests/integration/pilot/cni/testdata/cni.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+apiVersion: install.istio.io/v1alpha1
```
Comment (howardjohn) [inline tests/integration/pilot/cni/testdata/cni.yaml:1]: I think this is dead code

**Review by howardjohn — APPROVED** [review]
LGTM, left some comments that may speed things up since this tests adds 90s to our test runtime

File: tests/integration/pilot/cni/cni_race_test.go (hunk 5 above)
Comment (bianpengyuan) [reply tests/integration/pilot/cni/cni_race_test.go]: Move into the common suite.

File: tests/integration/pilot/cni/testdata/cni.yaml (hunk 7 above)
Comment (bianpengyuan) [reply tests/integration/pilot/cni/testdata/cni.yaml]: Removed. thanks!

**Pull request merged**